### PR TITLE
Feature NXP-22019 change token

### PR DESCRIPTION
**Pull request by efge**
For initial review. Tickets are:
- https://jira.nuxeo.com/browse/NXP-22259 (Internal document version/timestamp)
- https://jira.nuxeo.com/browse/NXP-22019 (Improve internal change token for optimistic locking)

### CODE REVIEW

File: nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/SessionImpl.java [hunk 1]
```diff
@@ -976,6 +979,10 @@ public Node addProxy(Serializable targetId, Serializable versionableId, Node par
         Node proxy = addChildNode(parent, name, pos, Model.PROXY_TYPE, false);
         proxy.setSimpleProperty(Model.PROXY_TARGET_PROP, targetId);
         proxy.setSimpleProperty(Model.PROXY_VERSIONABLE_PROP, versionableId);
+        if (changeTokenEnabled) {
+            proxy.setSimpleProperty(Model.MAIN_SYS_CHANGE_TOKEN_PROP, Model.INITIAL_SYS_CHANGE_TOKEN);
+            proxy.setSimpleProperty(Model.MAIN_CHANGE_TOKEN_PROP, Model.INITIAL_CHANGE_TOKEN);
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/SessionImpl.java]: Why do we set the change token property for a proxy ? Is it always created by a user ?

File: nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentLive.java [hunk 2]
```diff
@@ -338,6 +340,35 @@ public String getChangeToken() {
...
+            try {
+                modified = (Calendar) getPropertyValue(DC_MODIFIED);
+            } catch (PropertyNotFoundException e) {
+                modified = null;
+            }
+            if (modified == null) {
+                return true;
+            }
+            String currentToken = String.valueOf(modified.getTimeInMillis());
+            return userVisibleChangeToken.equals(currentToken);
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentLive.java]: This part of code doesn't seem to be present in DBS, is it on purpose ?

File: nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentProxy.java [hunk 3]
```diff
@@ -217,9 +217,60 @@ public void setSystemProp(String name, Serializable value) {
...
+            proxyToken = null;
+        }
+        String targetToken = parts[1];
+        if (targetToken.isEmpty()) {
+            targetToken = null;
+        }
+        if (proxyToken == null && targetToken == null) {
+            return true;
+        }
+        return proxy.validateUserVisibleChangeToken(proxyToken) && target.validateUserVisibleChangeToken(targetToken);
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentProxy.java]: `validateUserVisibleChangeToken` are more or less the same for DBS and VCS, can't we do anything for that ?

File: nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSTransactionState.java [hunk 4]
```diff
@@ -599,7 +617,15 @@ public void save() {
                 }
                 ChangeTokenUpdater changeTokenUpdater;
                 if (session.changeTokenEnabled) {
-                    changeTokenUpdater = new ChangeTokenUpdater(docState);
+                    // increment system change token
+                    Long base = (Long) docState.get(KEY_SYS_CHANGE_TOKEN);
```
Comment (efge) [inline nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSTransactionState.java]: True but it's followed by a `put(KEY_SYS_CHANGE_TOKEN, ...)` so for symmetry I'll keep it that way.

File: nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/SessionImpl.java (hunk 1 above)
Comment (efge) [inline nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/SessionImpl.java]: We initialize the values to 0 at creation time. Proxies can have user-modified proxy-specific schemas (in theory).

File: nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentLive.java (hunk 2 above)
Comment (efge) [inline nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentLive.java]: No you're right, I'm missing the legacy `dc:modified`-based change token code for DBS. I'll add it.

File: nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentProxy.java (hunk 3 above)
Comment (efge) [inline nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentProxy.java]: I tried but the way properties are accessed is different, and more importantly the way we treat proxy documents (flag for DBS vs separate class for VCS) makes it difficult. As it's not hugely complicated code I'll keep this.

**Comment by efge** [comment]
#947 is a new version of this pull request.

**Comment by efge** [comment]
#948 is the new PR for this.

**Pull request closed without merging**